launch/launch/utilities/type_utils.py

```diff
@@ -0,0 +1,451 @@
+    return output
+
+
+def is_substitution(x):
+    # is some substitution
+    return (
+        isinstance(x, Substitution) or
+        (
+            isinstance(x, collections.abc.Iterable) and
+            len(x) > 0 and
+            all(isinstance(item, (Substitution, str)) for item in x) and
+            any(isinstance(item, Substitution) for item in x)
+        )
+    )
+
+
+def normalize_typed_substitution(
+    value: SomeValueType, data_type: Optional[AllowedTypesType]
+) -> NormalizedValueType:
+    # Resolve scalar types immediately
+    if isinstance(value, ScalarTypesTuple):
+        if not check_type(value, data_type):
+            raise TypeError(f"value='{value}' is not an instance of {data_type}")
+        return value
+    # Resolve substitutions and list of substitutions immediately
+    if is_substitution(value):
+        return normalize_to_list_of_substitutions(cast(SomeSubstitutionsType, value))
+    # For the other cases, the input must be an iterable
+    if not isinstance(value, collections.abc.Iterable):
+        raise TypeError(
+            'value should be either a scalar, a substitutions,'
+            ' or a mixed list of scalars and substitutions. '
+            f'Got `value={value}` of type `{type(value)}`. '
+        )
+    # Collect the types of the items of the list
+    types_in_list: Set[Optional[Type[Union[str, int, float, bool, Substitution]]]] = set()
+    for x in value:
+        if isinstance(x, ScalarTypesTuple):
+            types_in_list.add(type(x))
+        elif is_substitution(x):
+            types_in_list.add(Substitution)
+        else:
+            types_in_list.add(None)
+    if None in types_in_list:
+        raise TypeError(
+            'value is a list, and one of the items is not a scalar, a Substitution '
+            f"or a list of substitutions. Got value='{value}'"
+        )
+    # Extract expected type information
+    is_list = True
+    if data_type is not None:
+        data_type, is_list = extract_type(data_type)
+    # Must be expecting a list
+    if not is_list:
+        raise TypeError(
+            'The provided value resolves to a list, though the required type is a scalar. '
+            f"Got value='{value}', data_type='{data_type}'."
+        )
+
+    # Normalize each specific uniform list input
+    err_msg = (
+        "Got a list of '{}'"
+        f", expected a list of '{data_type}'. value='{value}'"
+    )
+    if types_in_list.issubset({str, Substitution}):
+        # list of mixed strings and substitutions
+        if data_type not in (None, str):
+            raise TypeError(err_msg.format(str))
+        return cast(List[Union[List[Substitution], str]], [
+            x if isinstance(x, str) else  # Don't convert strings to TextSubstitution
+            normalize_to_list_of_substitutions(cast(SomeSubstitutionsType, x)) for x in value
+        ])
+    if types_in_list.issubset({bool, Substitution}):
```

Comment:
@ivanpauno if `types_in_list == {float, int, bool, Substitution}`, this will not fail until the "typed" substitution is performed. Is that intentional?

Reply:
Why?
It should fail right away.

Reply:
According to the tests I've added, that case is failing right away.

Reply:
Uh, I'm curious where and how it fails.